### PR TITLE
Allow assignment of attributes to multiple product types

### LINKED ISSUE #4252: Allow assigning an attribute to a product type
The attributes need to be available for assignment to a product type, and one attribute may be assigned to multiple product types.

- [x] Change the relationship between `Attribute` and `ProductType` from ForeignKey to M2M (both for `product_attributes` and `variant_attributes`)
- [x] add mutation to assign attributes: `assignAttributes(productTypeId, [attributeId])`
- [x] add mutation to unassign attribtues: `unassignAttributes(productTypeId, [attributeId])`

We might also need a query to return only attributes that are available for assignment for particular product type (e.g., I have five attributes in total, three are already assigned to the type that I'm editing, so I shouldn't be able to add them again). But maybe that's something that frontend will be able to handle on their side.

### PR DESCRIPTION
**Pull request by NyanKiyoshi**
Closes #4252.

### Definition of Done
- [x] Old product type attribute data is correctly migrated to the new relations
- [x] All the storefront and dashboard database queries were adapted for M2M usage
- [x] The dashboard product type & attribute sections are compatible with the new relations
- [x] The `assign` and `unassign` bulk mutations are implemented and tested as `productTypeId, [attributeId]`
  - [x] `assign(...)`
  - [x] `unassign(...)`
- [x] The `availableAttributes` connection field is implemented into `productType`
- [x] The changes are rebased and no merging between the product migrations was done

### Pull Request Checklist

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] Database migration files are up to date.
1. [x] The changes are tested.
1. [x] The code is documented (docstrings, project documentation).
1. [x] GraphQL schema and type definitions are up to date.
1. [x] Changes are mentioned in the changelog.

### CODE REVIEW

**Comment by NyanKiyoshi** [comment]
I changed the base to https://github.com/mirumee/saleor/pull/4260, otherwise the PR would have been blocked.

**Comment by django-queries** [comment]
Here is the report for e5e718a543fa8d9855b20ac9e9825b7d9d864664 (NyanKiyoshi/saleor @ attributes/gql/assign-m2m)
Base comparison is 7a94b1790ef6869982708c36186a71590f8e23df.



<details><summary>No differences were found. (click me)</summary>
<p>

```diff
# api.benchmark checkout
  test name                           	left count 	right count
  ------------------------------------	-----------	-----------
  add billing address to checkout     	         41	         41
  add shipping to checkout            	          7	          7
  checkout payment charge             	         16	         16
  complete checkout                   	          6	          6
  create checkout                     	         45	         45

# api.benchmark homepage
  test name                           	left count 	right count
  ------------------------------------	-----------	-----------
  retrieve main menu                  	          5	          5
  retrieve product list               	          4	          4
  retrieve secondary menu             	          5	          5
  retrieve shop                       	          2	          2

# api.benchmark product
  test name                           	left count 	right count
  ------------------------------------	-----------	-----------
  product details                     	         16	         16

# api.benchmark variant
  test name                           	left count 	right count
  ------------------------------------	-----------	-----------
  retrieve variant list               	         13	         13
```

</p>
</details>

**Comment by NyanKiyoshi** [comment]
Note: this PR depends on #4260. So #4260 needs to be merged first.

File: saleor/graphql/schema.graphql [hunk 1]
```diff
@@ -1509,9 +1523,11 @@ type Mutations {
   shippingZoneDelete(id: ID!): ShippingZoneDelete
   shippingZoneBulkDelete(ids: [ID]!): ShippingZoneBulkDelete
   shippingZoneUpdate(id: ID!, input: ShippingZoneInput!): ShippingZoneUpdate
-  attributeCreate(id: ID!, input: AttributeCreateInput!, type: AttributeTypeEnum!): AttributeCreate
+  attributeCreate(input: AttributeCreateInput!): AttributeCreate
   attributeDelete(id: ID!): AttributeDelete
   attributeBulkDelete(ids: [ID]!): AttributeBulkDelete
+  attributeAssign(operations: [AttributeAssignInput]!, productTypeId: ID!): AttributeAssign
+  attributeUnAssign(operations: [AttributeAssignInput]!, productTypeId: ID!): AttributeUnAssign
```
Comment (maarcingebala) [inline saleor/graphql/schema.graphql]: It should be called `attributeUnassign` and the return type should be `AttributeUnassign`.

**Pull request merged**